Type equivalence not inferred in function passed to second parameter list when first parameter list is a partial function

I've stumbled across a bafflingly specific case in which a type refined by a pattern won't be inferred as equivalent to the matched type. In the code:

```scala
object RefinementTest {
  trait A[T]
  trait B[T] extends A[T]

  def takesPartial
      (first: PartialFunction[Any, Unit])
      (second: () => Unit): Unit = ???

  def foo[T](a: A[T]) {
    a match {
      case b: B[t] =>
        takesPartial({ case _ => })(() => implicitly[t =:= T])
    }
  }
}
```

the implicit evidence is not found. All of the following factors are necessary to produce this bug. If any are removed, the problem goes away:

- The context is in the body of a lambda
- The lambda is passed as the second parameter list to a function taking two parameter lists
- The first parameter list is a partial function

Looks like this was fixed in 2.12 (tested with 2.12.8)?  Stays working in 2.13.0 too.

```bash
16:04:26 $ scalac -211 -version && scalac -211 RefinementTest.scala
Scala compiler version 2.11.12 -- Copyright 2002-2017, LAMP/EPFL
RefinementTest.scala:12: error: Cannot prove that t =:= T.
        takesPartial({ case _ => })(() => implicitly[t =:= T])
                                                    ^
one error found

16:04:38 ! scalac -212 -version && scalac -212 RefinementTest.scala
Scala compiler version 2.12.8 -- Copyright 2002-2018, LAMP/EPFL and Lightbend, Inc.

16:04:46 $ scalac -version && scalac RefinementTest.scala
Scala compiler version 2.13.0 -- Copyright 2002-2019, LAMP/EPFL and Lightbend, Inc.
```

Can we close this?  Should I submit the test-case and close this issue with that pull request?

--- a/test/files/pos/t9352.scala
+++ b/test/files/pos/t9352.scala
@@ -1,0 +1,15 @@
+object RefinementTest {
+  trait A[T]
+  trait B[T] extends A[T]
+
+  def takesPartial
+      (first: PartialFunction[Any, Unit])
+      (second: () => Unit): Unit = ???
+
+  def foo[T](a: A[T]) {
+    a match {
+      case b: B[t] =>
+        takesPartial({ case _ => })(() => implicitly[t =:= T])
+    }
+  }
+}

